x-pack/platform/plugins/shared/observability_ai_assistant/server/utils/rewrite_query.ts

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { findLast, pick } from 'lodash';
+import { BoundInferenceClient } from '@kbn/inference-plugin/server';
+import { Message, MessageRole } from '../../common';
+import { KnowledgeBaseQueryContainer } from '../service/knowledge_base_service';
+
+export async function rewriteQuery({
+  inferenceClient,
+  messages,
+  systemMessage,
+  context,
+}: {
+  inferenceClient: BoundInferenceClient;
+  messages: Message[];
+  systemMessage?: string;
+  context: string;
+}): Promise<{
+  queries: KnowledgeBaseQueryContainer[];
+}> {
+  const lastUserMessage = serialize(
+    findLast(
+      messages,
+      (message) => message.message.role === MessageRole.User && !!message.message.name
+    )
+  );
+
+  function serialize(message?: Message) {
+    return message
+      ? JSON.stringify(pick(message.message, 'content', 'name', 'function_call', 'role'))
+      : '';
+  }
+
+  const serializedConversation = messages.map((message) => {
+    return JSON.stringify(pick(message.message, 'content', 'name', 'function_call', 'role'));
+  });
+
+  const parts: string[] = [
+    `Below is a conversation from the user. Given
+      the information in the conversation, write queries that
+      can be executed against the user's knowledge base.
+
+      A semantic query is often a question, but it can be any
+      text that will be used for semantic searches.
+
+      Use keyword searches for named entities.
+
+      You can boost a query. A higher value for \`boost\` means
+      that a document that matches it will be scored higher.
+      
+      ### Example Prompt-to-Queries Rewrites
+
+      **Example 1**
+
+      User Prompt:  
+      "What can you tell me about my open alerts?"
+
+      Rewrite into Queries:  
+      - Semantic: "information on open alerts"  
+        Keywords: []  
+        Boost: 1  
+
+      - Semantic: "details and resolution steps for open alerts"  
+        Keywords: []  
+        Boost: 2  
+
+      ---
+
+      **Example 2**
+
+      User Prompt:  
+      "Visualize the count of my zookeeper logs over time."
+
+      Context:  
+      "A snippet in GitHub mentions 'zookeeper staging'."
```

Comment:
I know this is just an example but we wouldn't have retrieved entries from search connectors at this point, right? So we cannot know about anything on Github.

Perhaps better to make a realistic example that includes something from screen context.